docs/concepts/what-is-a-failure.md

```diff
@@ -0,0 +1,134 @@
+---
+id: what-is-a-failure
+title: What is a Failure?
+sidebar_label: Failure
+description: A Failure is Temporal's representation of various types of errors that occur in the system.
+tags:
+  - term
+  - explanation
+---
+
+A Failure is Temporal's representation of various types of errors that occur in the system.
+
+There are different types of Failures, and each has a different type in the SDKs and different information in the protobuf messages (which are used to communicate with the Cluster and appear in [Event History](/concepts/what-is-an-event-history)).
+
+## Temporal Failure
+
+Most SDKs have a base class that the other Failures extend:
+
+- TS: [`TemporalFailure`](https://typescript.temporal.io/api/classes/client.TemporalFailure)
+- Java: [`TemporalFailure`](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/failure/TemporalFailure.html)
+- Python: [`TemporalError`](https://python.temporal.io/temporalio.exceptions.TemporalError.html)
+
+The base [`Failure` proto message](https://github.com/temporalio/api/blob/e381e51864ec8f43a90750ef936705258b8f64b2/temporal/api/failure/v1/message.proto#L81-L114) has these fields:
+
+- `string message`
+- `string stack_trace`
+- `string source`: The SDK this Failure originated in (for example, `"TypeScriptSDK"`). In some SDKs, this field is used to rehydrate the stack trace into an exception object.
+- `Failure cause`: The `Failure` message of the cause of this Failure (if applicable).
+
+## Application Failure
+
+Workflow and Activity code use Application Failures to communicate application-specific failures that happen.
+This is the only type of Failure created and thrown by user code.
+
+- TS: [`ApplicationFailure`](https://typescript.temporal.io/api/classes/client.ApplicationFailure)
+- Java: [`ApplicationFailure`](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/failure/ApplicationFailure.html)
+- Go: [`ApplicationError`](https://pkg.go.dev/go.temporal.io/sdk/internal#ApplicationError)
+- Python: [`ApplicationError`](https://python.temporal.io/temporalio.exceptions.ApplicationError.html)
+- Proto: [`ApplicationFailureInfo`](https://github.com/temporalio/api/blob/e381e51864ec8f43a90750ef936705258b8f64b2/temporal/api/failure/v1/message.proto#L37-L41) and [`Failure`](https://github.com/temporalio/api/blob/e381e51864ec8f43a90750ef936705258b8f64b2/temporal/api/failure/v1/message.proto#L81-L114)
+
+### Throw from Workflows
+
+In Workflows, if you throw (or return in Go) an error that is not an Application Failure, the Workflow Task will fail and be retried.
+If you throw an Application Failure, the Workflow Execution will fail, and may be retried according to its Retry Policy.
```

Comment:
Only Workflow errors that are Temporal Failures will cause the Worklow Execution to fail: all other errors will cause the Workflow Task to fail and be retried.
Most types of Temporal Failures automatically occur, like a [Cancelled Failure](#cancelled-failure) when the Workflow is Cancelled or an [Activity Failure](#activity-failure) when an Activity Fails.
You can also explicitly fail the Workflow Execution by throwing (or returning, depending on the SDK) an Application Failure.

Reply:
Sounds great! This is a hugely important concept.

I should add that I just remembered I don't think this isn't true for Go :-) Since errors are explicitly returned, I think we treat _every_ workflow error as workflow execution failure (but a panic as workflow task failure).